src/net/ossl.cc

```diff
@@ -173,4 +188,262 @@ tls::dh_params::~dh_params() {
+        }
+        *cert = x509_cert;
+        return *cert;
+    }
+
+    void set_x509_trust(const blob& b, x509_crt_format fmt) {
+        auto store_ctx = x509_store_ctx_ptr(X509_STORE_CTX_new());
+        X509* x509_cert = nullptr;
+        if(nullptr == parse_x509_cert(b, fmt, &x509_cert)){
+            throw ossl_error();
+        }
+        X509_STORE_add_cert(*this, x509_cert);
+        _certs.push_back(x509_ptr(x509_cert));
+    }
+
+    void set_x509_crl(const blob& b, x509_crt_format fmt) {
+        auto store_ctx = x509_store_ctx_ptr(X509_STORE_CTX_new());
+        bio_ptr cert_bio(BIO_new_mem_buf(b.begin(), b.size()));
+        X509_CRL* x509_crl = X509_CRL_new();
+        if(fmt == x509_crt_format::PEM) {
+            if (nullptr == PEM_read_bio_X509_CRL(cert_bio.get(), &x509_crl, nullptr, nullptr)){
+                X509_CRL_free(x509_crl);
+                throw ossl_error();
+            }
+        } else if (fmt == x509_crt_format::DER){
+            if (nullptr == d2i_X509_CRL_bio(cert_bio.get(), &x509_crl)){
+                X509_CRL_free(x509_crl);
+                throw ossl_error();
+            }
+        } else {
+            throw std::invalid_argument("Unsupported cert format");
+        }
+        X509_STORE_add_crl(*this, x509_crl);
+        _crls.push_back(x509_crl_ptr(x509_crl));
+    }
+
+    void set_x509_key(const blob& cert, const blob& key, x509_crt_format fmt) {
+        // Theres no interface to add cert-key pair to the certificate store as
+        // the store represents the root and intermediate chain. Exposed for later
+        // use when the ssl socket is created
+        X509* x509_tmp = nullptr;
+        if(nullptr == parse_x509_cert(cert, fmt, &x509_tmp)){
+            throw ossl_error();
+        }
+        auto x509_cert = x509_ptr(x509_tmp);
+        bio_ptr key_bio(BIO_new_mem_buf(key.begin(), key.size()));
+        auto pkey_temp = EVP_PKEY_new();
+        if (nullptr == PEM_read_bio_PrivateKey(key_bio.get(), &pkey_temp, nullptr, nullptr)) {
+            EVP_PKEY_free(pkey_temp);
+            throw ossl_error();
+        }
+        auto pkey = evp_pkey_ptr(pkey_temp);
+        _ck_pair = std::make_optional(cert_key_pair{.cert = std::move(x509_cert), .key = std::move(pkey)});
+    }
+
+    void set_simple_pkcs12(const blob& b, x509_crt_format fmt, const sstring& password) {
+        // Load the PKCS12 file
+        bio_ptr bio(BIO_new_mem_buf(b.begin(), b.size()));
+        PKCS12 *p12_tmp = nullptr;
+        if(nullptr == d2i_PKCS12_bio(bio.get(), &p12_tmp)) {
+            throw ossl_error();
+        }
+        auto p12 = pkcs12(p12_tmp);
+        // Extract the certificate and private key from PKCS12, using provided password
+        EVP_PKEY *pkey = nullptr;
+        X509 *cert = nullptr;
+        if (!PKCS12_parse(p12.get(), password.c_str(), &pkey, &cert, nullptr)) {
+            throw ossl_error();
+        }
+        X509_STORE_add_cert(*this, cert);
+        EVP_PKEY_free(pkey);
+    }
```

Comment:
I think you'll want to store the key pair like you're doing above and not throw away the key.  This is just another way of setting the cert and paired key.

Also, the PKCS12 may also contain a certificate chain that you should also add to the store.